promql/test.go

```diff
@@ -0,0 +1,506 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promql
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	clientmodel "github.com/prometheus/client_golang/model"
+
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/local"
+	"github.com/prometheus/prometheus/storage/metric"
+	"github.com/prometheus/prometheus/utility"
+	testutil "github.com/prometheus/prometheus/utility/test"
+)
+
+var (
+	minNormal = math.Float64frombits(0x0010000000000000) // The smallest positive normal value of type float64.
+
+	patSpace       = regexp.MustCompile("[\t ]+")
+	patLoad        = regexp.MustCompile(`^load\s+(.+?)$`)
+	patEvalInstant = regexp.MustCompile(`^eval(?:_(fail|ordered))?\s+instant\s+(?:at\s+(.+?))?\s+(.+)$`)
+)
+
+const (
+	testStartTime = clientmodel.Timestamp(0)
+	epsilon       = 0.000001 // Relative error allowed for sample values.
+	maxErrorCount = 10
+)
+
+// Test is a sequence of read and write commands that are run
+// against a test storage.
+type Test struct {
+	*testing.T
+
+	cmds []testCommand
+
+	storage      local.Storage
+	closeStorage func()
+	queryEngine  *Engine
+}
+
+// NewTest returns an initialized empty Test.
+func NewTest(t *testing.T, input string) (*Test, error) {
+	test := &Test{
+		T:    t,
+		cmds: make([]testCommand, 0),
```

Comment:
`cmds` doesn't need to be initialized (and if it did, `[]testCommand{}` is our usual style for initialization (only use `make` when no other syntax exists))